### PR TITLE
o/snapstate: respect validation sets when refreshing and installing components with snaps

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This change makes it so that we respect validation sets when installing/refreshing components alongside snaps.

Note that the store doesn't yet fully support introspecting the validation sets that are sent during a refresh, so they cannot yet change which component revisions are actually returned with the actions. We double check the actions that are returned from the store to ensure that we always follow the correct constraints.

Based on https://github.com/canonical/snapd/pull/14769 for now.

### CODE REVIEW

File: overlord/snapstate/target.go [hunk 1]
```diff
@@ -312,11 +312,103 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+		if action == "refresh" {
+			verb = "update"
+		}
+
+		return fmt.Errorf(
+			"cannot %s snap %q due to enforcing rules of validation set %s",
+			verb,
+			instanceName,
+			constraints.Sets.CommaSeparated(),
+		)
```
Comment (miguelpires) [inline overlord/snapstate/target.go:366]: Sorry to be nitpicking on style but to me this more verbose and actually less readable than just splitting it in two lines
```suggestion
		return fmt.Errorf("cannot %s snap %q due to enforcing rules of validation set %s",
			verb, instanceName, constraints.Sets.CommaSeparated())
```
IMO it makes it easier to match the format specifiers to the variables

File: overlord/snapstate/target.go [hunk 2]
```diff
@@ -312,11 +312,103 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+
+// should something like this maybe be a method on snapasserts.ValidationSets?
+// the error strings are the main thing that don't really fit well in there
+func checkSnapAgainstConstraints(
+	instanceName string,
+	revision snap.Revision,
+	constraints snapasserts.SnapPresenceConstraints,
+	action string,
+) error {
+	if constraints.Presence == asserts.PresenceInvalid {
```
Comment (miguelpires) [inline overlord/snapstate/target.go:355]: I think we're missing a test for this case

File: overlord/snapstate/target.go [hunk 3]
```diff
@@ -312,11 +312,103 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+	comps := make(map[string]snap.Revision, len(target.components))
+	for _, comp := range target.components {
+		comps[comp.ComponentName()] = comp.Revision()
+	}
+
+	return checkComponentsAgainstConstraints(target.info.SnapName(), comps, constraints, action)
+}
+
+// should something like this maybe be a method on snapasserts.ValidationSets?
+// the error strings are the main thing that don't really fit well in there
```
Comment (miguelpires) [inline overlord/snapstate/target.go:348]: It seems fine as a standalone function, a method doesn't feel particularly more natural than this

**Review by miguelpires — COMMENTED** [review]
looks good but missing a test I think

File: overlord/snapstate/target.go (hunk 2 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:353]: Fixed, thanks. There was an identical check right before calling this function, preventing the test that should have hit it from going in there.

File: overlord/snapstate/target.go [hunk 4]
```diff
@@ -312,11 +312,97 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+	constraints snapasserts.SnapPresenceConstraints,
+	action string,
+) error {
+	if constraints.Presence == asserts.PresenceInvalid {
+		verb := "install"
+		if action == "refresh" {
+			verb = "update"
+		}
+
+		return fmt.Errorf("cannot %s snap %q due to enforcing rules of validation set %s",
```
Comment (pedronis) [inline overlord/snapstate/target.go:359]: shouldn't this error mention "invalid presence" somehow?

File: overlord/snapstate/target.go [hunk 5]
```diff
@@ -312,11 +312,97 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+	if action == "refresh" {
+		verb = "update"
+	}
+
+	for compName, compRevision := range comps {
+		cp := constraints.Component(compName)
+
+		if cp.Presence == asserts.PresenceInvalid {
+			return fmt.Errorf(
+				"cannot %s component %q due to enforcing rules of validation set %s",
```
Comment (pedronis) [inline overlord/snapstate/target.go:381]: same for this one

File: overlord/snapstate/target.go [hunk 6]
```diff
@@ -312,11 +312,97 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
 		})
 	}
 
-	// TODO: check validation sets here to catch invalid component revisions
+	for _, t := range installs {
+		sn, ok := s.find(t.info.InstanceName())
```
Comment (pedronis) [inline overlord/snapstate/target.go:316]: I don't remember this method, wouldn't "get" be a better name for it?

File: overlord/snapstate/snapmgr.go [hunk 7]
```diff
@@ -699,6 +699,24 @@ func (snapst *SnapState) InstanceName() string {
 	return snap.InstanceName(cur.RealName, snapst.InstanceKey)
 }
 
+// SnapName returns the name of the snap. Implementation of naming.SnapRef.
+func (snapst *SnapState) SnapName() string {
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:703]: maybe it's obvious but where is this used?

File: overlord/snapstate/snapmgr.go [hunk 8]
```diff
@@ -699,6 +699,24 @@ func (snapst *SnapState) InstanceName() string {
...
+	cur := snapst.CurrentSideInfo()
+	if cur == nil {
+		return ""
+	}
+	return cur.RealName
+}
+
+// ID returns the ID of the snap. Implementation of naming.SnapRef.
+func (snapst *SnapState) ID() string {
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:712]: same question

**Review by pedronis — COMMENTED** [review]
did a pass, some comments/questions

**Review by miguelpires — APPROVED** [review]
thank you

File: overlord/snapstate/target.go (hunk 4 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:359]: These error strings aren't new, they are present in some of our spread tests too. I'm happy to change them, though if you think we should?

File: overlord/snapstate/snapmgr.go (hunk 7 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapmgr.go:703]: Hrm, I guess I factored the need for this out at some point. Reverting those commits.

File: overlord/snapstate/target.go (hunk 6 above)
Comment (pedronis) [reply overlord/snapstate/target.go:316]: @andrewphelpsj maybe even snap() ?

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**